### PR TITLE
Cleanups and new features

### PR DESCRIPTION
**Pull request by jleclanche**
WIP; will be ready soon.

### CODE REVIEW

**Comment by jleclanche** [comment]
You can start reviewing if you like. I'm testing this in production at the same time FYI so I'm fixing and implementing things as I find them.

**Comment by samamorgan** [comment]
Can you get this over to the develop branch instead?

**Comment by samamorgan** [comment]
I'm working on merging this into develop, you may want to hold off on commits then pull develop and start pushing there instead. I'll work towards another dot release when this is ready!

**Comment by samamorgan** [comment]
I intentionally left out .gitignore changes, I want that to be as catch-all as possible to prevent accidental file commits on other PRs

**Comment by jleclanche** [comment]
Thanks! I'm still testing this but everything appears to work. I'll file another PR if I have more stuff.

**Comment by samamorgan** [comment]
Responding here to comments you've made on recent commits to close the loop.

> 7d09a07 I kinda wish you'd have discussed this in the PR, there's other ways to keep the library user-friendly without having the objects be stateful.

I reverted the decision to make objects "stateless" (they weren't stateless from my understanding, they're still initialized in the super class). One of the primary design goals I have with this lib is to make everything as automatic for users as possible. I don't see an advantage to forcing users to hang on to a client object to make repeated calls to a JsonObject child.

> c0f78ae Same here. Why do you need vars()? You can just inspect .json.

I also reverted the changes to JsonObject. In my opinion, Python built-ins should be properly supported whenever possible. A user should be able to be passed an object and inspect it without much fuss. Hiding instance variables behind another variable makes the object less explicit, and again I don't see an advantage to that.

None of these decisions are set in stone, that's the purpose of the dev branch, and I'm glad to discuss further changes. What were your goals behind these design decisions?

**Comment by jleclanche** [comment]
Thanks for the answers -- I hope I didn't come across as rude, I was just surprised you reverted them instead of discussing here directly before merging.

> they weren't stateless from my understanding, they're still initialized in the super class

Stateless as in, can be initialized from just the JSON data. This IMO is important for cases where you need to serialize this (using eg. `json.dumps(obj._data)`) and then reserialize it (`Group(data=json.loads(data))`). With your approach, you have to have an API client at the ready, which is irrelevant unless you actually want to operate on these objects…

I dunno, it might not make that much of a difference in the end. A good reference for sdk design is the Stripe Python API, which doesn't even use instanciated classes for most of the work:

https://github.com/stripe/stripe-python/blob/master/stripe/

> In my opinion, Python built-ins should be properly supported whenever possible. A user should be able to be passed an object and inspect it without much fuss. Hiding instance variables behind another variable makes the object less explicit, and again I don't see an advantage to that.

This makes the library unusable for me FWIW so I've had to go back to my own fork. I need to be able to reference the `_data` dict itself because I may subclass the objects, which will break any `__dict__` reference expecting only json-serializable objects, unless I double-underscore all the methods. In general, using `vars()` and/or `__dict__` in any non-debug code is a code smell for that very reason.

If you're okay changing this back I can work on resyncing my fork and PRing my two latest commits as well:

- https://github.com/jleclanche/discourse/commit/b10c98c8500bdd30db9740e2faac6b2f0f288569
- https://github.com/jleclanche/discourse/commit/eb2c937f32ef2229dfcc48db7d82608da7275dfd

**Comment by samamorgan** [comment]
Thanks for re-engaging! No problem on the response tone, I understand how frustrating it can be when work is seemingly wasted. I absolutely don't want anything set in stone with this library this early in development, so don't fret too much!

I'm not sure if you noticed, but I did add JsonObject.json() to address the serialization issue. This allows you to simply call object.json() to get back pure serializable data. I think having a helper function like this is important, as I do intend for most of these objects to have sub-objects as well, for instance Topics will contain Post objects. JsonObject.json() would need to iterate over self.__dict__ and call json() on sub-objects to get serializable-only data.

https://github.com/samamorgan/discourse/blob/develop/discourse/jsonobject.py#L24

Regarding subclassing, I'm not fully understanding what you're trying to convey here. Can you show some code example(s) that explain? IMO code is usually better than words!

One very large disadvantage I see to not passing client objects to the JsonObject initializer is this:

### With Client passed:

**a.py**
```python
def first_post(topic):
    first = self.post_stream['posts'][0]
    return topic.get_post(first['id'])
```
**b.py**
```python
import a

client = discourse.Client(host, auth)
topic = client.get_topic(12345)

first_post = a.first_post(topic)
```

### Without Client passed:
**a.py**
```python
def first_post(topic, client):
    first = self.post_stream['posts'][0]
    return client.get_post(first['id'])
```
**b.py**
```python
import a

client = discourse.Client(host, auth)
topic = client.get_topic(12345)

first_post = a.first_post(topic, client)
```

This pattern applies in single file scripts as well, except in that case you can just define the client in global scope, which is a bad convention for many reasons. Forcing users to pass a client would often force this bad coding practice.

Additionally, the second example is fundamentally different. This would force another request since we would be grabbing from the API rather than building from the already instantiated Topic.

**Comment by samamorgan** [comment]
The above isn't a great example, I see a few flaws in the reasoning, but I hope you understand the overall reasoning.

**Comment by jleclanche** [comment]
> Regarding subclassing, I'm not fully understanding what you're trying to convey here. Can you show some code example(s) that explain?

The problem with your approach is that it sets everything in the json dict as an attribute, right? So now, if I do `instance.edit`, can you tell, am I accessing an attribute or a function?

With my approach, I am always accessing a function *first*, then an attribute if one doesn't exist. If I know I want an arbitrary attribute, I can access it with `instance._json["edit"]`. This is a much saner way to work with the code especially when dealing with arbitrary attributes. As I've gained experience with Python, I've learned to stay far away from its more dynamic elements :)

> Additionally, the second example is fundamentally different. This would force another request since we would be grabbing from the API rather than building from the already instantiated Topic.

I'm not seeing how the two are different. `topic` is still a topic instance with all its data, it's just that it doesn't have a `client` attribute so you have to make sure to pass it in.

Here's an example of something you can't do (cleanly) with your approach, but can with stateless objects, given the methods for it:

```py
forum1 = Client(apikey1)
forum2 = Client(apikey2)

post = forum1.get_some_post()
topic = forum2.get_some_topic()
topic.reply(post, client=forum2)
```

In other words, transferring data from one discourse instance to another is less clean with your approach because you can't instanciate objects.

Anyway, I think it's a better approach mostly because I've gravitated towards being less stateful over the years. But it's not a dealbreaker for me, so do as you prefer, it's just my recommendation :)
